enzyme/Enzyme/MLIR/Analysis/DataFlowActivityAnalysis.cpp

```diff
@@ -0,0 +1,976 @@
+    processFunc(curr);
+
+    curr.walk([&](CallOpInterface call) {
+      auto neighbor = dyn_cast_if_present<FunctionOpInterface>(
+          call.resolveCallable(symbolTable));
+      if (neighbor && !visited.contains(neighbor)) {
+        frontier.push_back(neighbor);
+        visited.insert(neighbor);
+      }
+    });
+  }
+}
+
+void printActivityAnalysisResults(const DataFlowSolver &solver,
+                                  FunctionOpInterface callee,
+                                  const SmallPtrSet<Operation *, 2> &returnOps,
+                                  SymbolTableCollection *symbolTable,
+                                  bool verbose, bool annotate) {
+  auto isActiveData = [&](Value value) {
+    auto fva = solver.lookupState<ForwardValueActivity>(value);
+    auto bva = solver.lookupState<BackwardValueActivity>(value);
+    bool forwardActive = fva && fva->getValue().isActiveVal();
+    bool backwardActive = bva && bva->getValue().isActiveVal();
+    return forwardActive && backwardActive;
+  };
+
+  auto isConstantValue = [&](Value value) {
+    // TODO: integers/vectors that might be pointers
+    if (isa<LLVM::LLVMPointerType, MemRefType>(value.getType())) {
+      assert(returnOps.size() == 1);
+      auto *fma = solver.lookupState<ForwardMemoryActivity>(*returnOps.begin());
+      auto *bma = solver.lookupState<BackwardMemoryActivity>(
+          &callee.getFunctionBody().front().front());
+
+      auto *pointsToSets =
+          solver.lookupState<enzyme::PointsToSets>(*returnOps.begin());
+      auto *aliasClassLattice = solver.lookupState<AliasClassLattice>(value);
+      // Traverse the points-to sets in a simple BFS
+      std::deque<DistinctAttr> frontier;
+      DenseSet<DistinctAttr> visited;
+      const DenseSet<DistinctAttr> &aliasClasses =
+          aliasClassLattice->getAliasClasses();
+      frontier.insert(frontier.end(), aliasClasses.begin(), aliasClasses.end());
+      visited.insert(aliasClasses.begin(), aliasClasses.end());
+      while (!frontier.empty()) {
+        DistinctAttr aliasClass = frontier.front();
+        frontier.pop_front();
+
+        // It's an active pointer if active data flows in from the forward
+        // direction and out from the backward direction.
+        if (fma->hasActiveData(aliasClass) &&
+            bma->activeDataFlowsOut(aliasClass))
+          return false;
+
+        // Or if it points to a pointer that points to active data
+        const enzyme::AliasClassSet &neighbors =
+            pointsToSets->pointsTo.lookup(aliasClass);
+        for (DistinctAttr neighbor : neighbors.getAliasClasses()) {
+          if (!visited.contains(neighbor)) {
+            visited.insert(neighbor);
+            frontier.push_back(neighbor);
+          }
+        }
+      }
+      // Otherwise, it's constant
+      return true;
+    }
+
+    return !isActiveData(value);
+  };
+
+  function_ref<bool(Operation *)> isConstantInstruction = [&](Operation *op) {
```

Comment:
(+) `function_ref` is a non-owning reference to a function. You most likely want an `std::function` here so the value is not a reference to a temporary.